Indicate an ended target process in new beta UI

beckerhe opened the pull request:
Since the process list was removed in the new (beta) UI it is
particurlary hard to realize that the target process has ended on the
instance.

This commit is adding a somewhat hacky fix. It's checking the regularly
incoming process list for existence of the target process. If the target
process pid cannot be found the corner widget will color the process
name in orange - indicating that the process has died. Additionally it
will disable the Start-Capture button.

Bug: http://b/176960277

reichlfl commented on OrbitQt/orbitmainwindow.cpp:
```diff
@@ -582,6 +582,13 @@ void OrbitMainWindow::UpdateCaptureStateDependentWidgets() {
 
+  // The detection mechanism is only implemented for the new UI, so we maintain
+  // the old behaviour for the old UI and assume the target process is always running.
+  // This can be renamed to is_target_process_running when the feature flag gets removed.
```
IIRC we have a bug id mentioned in all other places concerning removal of the feature flag, can you add this here as well with a TODO?

beckerhe replied:
Sure. I can add that.

My reasoning for not doing that: Whenever the code involves the feature flag `FLAGS_enable_ui_beta`, it's an even stronger indication that this has to be touched when the feature flag is removed. The compiler would complain if it was forgotten.

beckerhe replied:
done.

reichlfl commented on OrbitQt/orbitmainwindow.cpp:
```diff
@@ -1165,11 +1173,38 @@ void OrbitMainWindow::SetTarget(const orbit_qt::StadiaTarget& target) {
                          target.GetConnection()->GetInstance().display_name);
+
+  using ProcessInfo = orbit_grpc_protos::ProcessInfo;
+  target.GetProcessManager()->SetProcessListUpdateListener([&](std::vector<ProcessInfo> processes) {
```
For readability and to allow easier modification later, I'd prefer to provide a `OrbitMainWindow::OnProcessListUpdated` and call this inside `invokeMethod`.

beckerhe replied:
Makes sense. I debated that with myself as well!

beckerhe replied:
Done.

reichlfl commented on OrbitQt/orbitmainwindow.cpp:
```diff
@@ -1165,11 +1173,38 @@ void OrbitMainWindow::SetTarget(const orbit_qt::StadiaTarget& target) {
+    // to execute our logic on the main thread.
+    QMetaObject::invokeMethod(this, [&, processes = std::move(processes)]() {
+      const auto is_current_process = [this](const auto& process) {
+        const auto target_process = app_->GetTargetProcess();
```
I have been scolded for using `auto` too much, so I'll do the same here :-) Same for the `auto`s following below
https://google.github.io/styleguide/cppguide.html#Type_deduction

beckerhe replied:
Will change that! :+1:

beckerhe replied:
done. I just like auto too much. :see_no_evil: Time to move on to C++20!

reichlfl commented on OrbitQt/orbitmainwindow.cpp:
```diff
@@ -1165,11 +1173,38 @@ void OrbitMainWindow::SetTarget(const orbit_qt::StadiaTarget& target) {
+      const auto current_process =
+          std::find_if(processes.begin(), processes.end(), is_current_process);
+
+      if (const auto process_died = current_process == processes.end(); process_died) {
```
I don't see the benefit of the `process_died` assignment here since it's not used inside the if-block. Is it purely to convey intent of `current_process == processes.end()`? For me, it's more confusing then helping because I've never used this if-syntax in the past, but maybe that's just me.

In any case, `auto` instead of `bool` is certainly not helping ;-)

beckerhe replied:
Yes, it's merely to explain that `current_process == processes.end()` means that the process died. But I can write that differently.

beckerhe replied:
Changed!

reichlfl reviewed (commented):
Thanks for fixing this so quickly, as discussed I think it's perfectly fine to be somewhat of a workaround for now. Few comments mostly regarding coding style, functionality looks fine to me!

beckerhe commented:
Thanks Florian. I addressed your comments.

I had to make some changes to the stylesheet handling when I factored out this code into its own function. I'm not sure if I like it that way. Can you have another look?

I also realized, that we can enable the detection also for the local target. So I added that as well.

reichlfl commented on OrbitQt/orbitmainwindow.cpp:
```diff
@@ -1242,3 +1236,24 @@ void OrbitMainWindow::SetupGrpcAndProcessManager(std::string grpc_server_address
+    const ProcessData* const target_process = app_->GetTargetProcess();
+    return target_process != nullptr && process.pid() == app_->GetTargetProcess()->pid();
+  };
+  const auto current_process = std::find_if(processes.begin(), processes.end(), is_current_process);
```
Another `auto` that's probably not conforming to the style guide

beckerhe replied:
I'm not sure if `const std::vector<ProcessInfo>::const_iterator` helps readability here. Actually I believe it rather harms readability. WDYT?

reichlfl replied:
Oh sorry, I misread and thought this was supposed to be `ProcessData*`. My bad, ignore my comment.

beckerhe replied:
No worries at all. Glad that we agree on this. ;-)

beckerhe replied:
and FWIW. `std::vector<ProcessInfo>::const_iterator` might be an alias for `const ProcessInfo*` depending on the implemenation of the standard library. But we can't assume that.

reichlfl commented on OrbitQt/orbitmainwindow.h:
```diff
@@ -95,6 +95,13 @@ class OrbitMainWindow : public QMainWindow {
   // TODO(170468590): [ui beta] When out of ui beta, this can return TargetConfiguration (without
   // std::optional)
   std::optional<orbit_qt::TargetConfiguration> ClearTargetConfiguration() {
```
I've just realized this is in the header file - suggest to move the implementation to the cpp file.

beckerhe replied:
Yep. Good point. Done.

reichlfl reviewed (approved):
> Thanks Florian. I addressed your comments.
> 
> I had to make some changes to the stylesheet handling when I factored out this code into its own function. I'm not sure if I like it that way. Can you have another look?
> 
> I also realized, that we can enable the detection also for the local target. So I added that as well.

Thanks, I've only got small comments, but overall LGTM!

I'm not sure what exactly you're referring to here, you mean replacing the text with constants and pulling them out of the methods? It's still not ideal that we always define our colors locally (by locally, I don't mean only as a local variable in a function, but also across several cpp files and not in one central place), but it's better than before. We're doing this all across Orbit, so it's nothing we should improve in this PR now.
